include/xtensor/xpad.hpp

```diff
@@ -81,82 +81,104 @@ namespace xt
+        xt::xstrided_slice_vector sv;
         for (size_type axis = 0; axis < e.shape().size(); ++axis)
         {
             size_type nb = static_cast<size_type>(pad_width[axis][0]);
             size_type ne = static_cast<size_type>(pad_width[axis][1]);
+            size_type ns = nb + e.shape(axis) + ne;
+            new_shape[axis] = ns;
+            sv.push_back(xt::range(nb, nb + e.shape(axis)));
+        }
 
-            if (nb == size_type(0) && ne == size_type(0))
-            {
-                continue;
-            }
+        if (mode == pad_mode::constant)
+        {
+            return_type out(new_shape, constant_value);
+            xt::strided_view(out, sv) = e;
+            return out;
+        }
+
+        return_type out(new_shape);
+        xt::strided_view(out, sv) = e;
+
+        // construct padded regions based on original image
+
+        for (size_type axis = 0; axis < e.shape().size(); ++axis)
+        {
+            size_type nb = static_cast<size_type>(pad_width[axis][0]);
+            size_type ne = static_cast<size_type>(pad_width[axis][1]);
 
-            if (mode == pad_mode::constant)
+            if (nb > static_cast<size_type>(0))
             {
-                auto shape_bgn = out.shape();
-                auto shape_end = out.shape();
-                shape_bgn[axis] = nb;
-                shape_end[axis] = ne;
-                auto bgn = constant_value * xt::ones<value_type>(shape_bgn);
-                auto end = constant_value * xt::ones<value_type>(shape_end);
-                out = xt::concatenate(xt::xtuple(bgn, out, end), axis);
+                xt::xstrided_slice_vector svs(e.shape().size(), xt::all());
+                xt::xstrided_slice_vector svt(e.shape().size(), xt::all());
+
+                svt[axis] = xt::range(0, nb);
+
+                if (mode == pad_mode::wrap || mode == pad_mode::periodic)
+                {
+                    XTENSOR_ASSERT(nb <= e.shape(axis));
+                    svs[axis] = xt::range(e.shape(axis), nb+e.shape(axis));
+                }
+                else if (mode == pad_mode::symmetric)
+                {
+                    XTENSOR_ASSERT(nb <= e.shape(axis));
+                    svs[axis] = xt::range(2*nb-1, nb-1, -1);
+                }
+                else if (mode == pad_mode::reflect)
+                {
+                    XTENSOR_ASSERT(nb <= e.shape(axis) - 1);
+                    svs[axis] = xt::range(2*nb, nb, -1);
+                }
+
+                xt::strided_view(out, svt) = xt::strided_view(out, svs);
             }
-            else
+
+            if (ne > static_cast<size_type>(0))
             {
-                xt::xstrided_slice_vector sv_bgn(e.shape().size(), xt::all());
-                xt::xstrided_slice_vector sv_end(e.shape().size(), xt::all());
+                xt::xstrided_slice_vector svs(e.shape().size(), xt::all());
+                xt::xstrided_slice_vector svt(e.shape().size(), xt::all());
```

Comment:
We should reuse the vectors allocated out of the loop here. The slices modified during the axis-th iteration should be reset to `xt::all` before the next iteration.